# Shard_and_Jones_J_Neurooncol_2025
This repository contains R and MATLAB pipelines for single cell and spatial transcriptomics analysis associated with Shard and Jones et al., 2025 J Neurooncol

All pipelines for scRNAseq analysis are in "Shard_and_Jones_2025_Single_Cell_RNAseq_Rscript" - please refer to supplementary methods in Shard, Jones, et al., J_Neurooncol 2025. This code uses publicly available scRNA-seq data available through the GEO repository with the following dataset ID's: GSE173280, GSE182109, GSE141383, GSE184357 and GSE102130, as well as GBM scRNAseq data from Ebert et al. 2020, (available upon request from the authors). Cell type annotations were provided for all datasets by authors, except for GSE141383, which we used CellKb (https://www.cellkb.com) to annotate using glioma reference datasets. To annotate Neftel cell states use the "calculate_Neftel_states.Rmd"

All pipelines for spatial transcriptomics analysis are in "Shard_and_Jones_2025_Spatial_Transcriptomics_Rscript" - please refer to supplementary methods in Shard, Jones, et al., J_Neurooncol 2025. Spatial transcriptomics data from DMG samples were obtained from Ren et al., (2023) (GSE194329). GBM spatial transcriptomics data generated by this study will be available European genome-phenome (EGA) archive (available upon request from the authors).

MATLAB scripts are provided to perform Manders' coefficient analysis. This is compatible with the excel input file generated using the "Shard_and_Jones_2025_Spatial_Transcriptomics_Rscript". Perform scale correction first before Manders' coefficient analysis. 

